### PR TITLE
WIP: Static cache implementation

### LINKED ISSUE #157: Caching not working (as well as it could?) when using docker...
I don't know if this problem is Docker-specific or not, but the subject says it all... for local development cycles we use Macs, but our CI/CD servers deploy via Linux.  In order for things to work properly for our devs I've got numerous deployments of this plugin, but we've noticed the ever-increasing time of the requirements step.  The docker log shows that it uses cache, it appears this is only for the caching of the downloaded modules from the web, before they are setup.

What I'd like is if the requirements.txt file has not changed (md5 sum?) to not go and attempt to download and re-setup the requirements, ideally not even launching docker, since docker will do the same thing over and over.  In this case, all the plugin should do is re-symlink/cope the contents in place from the requirements cache folder, then zip up, and then remove those symlinks/contents.

I don't know if this is a bug, or a feature request.  But when I hear that this library is "caching" requirements, I assume it's caching more than just the downloading of packages, it should be caching the "output" of the requirements step, not the "prefetch" of the requirements step.

I'd love some feedback/input/thoughts from some other contributors/authors of this plugin.  Because I wouldn't mind doing this feature myself, but I want some input before diving into it.

### PR DESCRIPTION
**Pull request by AndrewFarley**
Fixes #157 (filed by me)

## What this does
This implements a static cache for the pip **outputs** (not to be confused with the downloaded pip modules which this module technically already does).  This greatly speeds up every single build as long as you have the feature enabled and do not change your requirements.txt file.  Since this file rarely changes, you'll see some amazing speed increases when packaging your service.  This _especially_ helps on machines that exclusively use Docker (Win/Mac) for deployments.

## Implementation details
* When enabled, this plugin now caches all requirements outputs to an "appdir" cache folder (per the [appdirectory](https://www.npmjs.com/package/appdirectory) node module).
* Within this new cache directory is a subdirectory with the folder name of the MD5 sum of the requirements file that was used to generate those requirements.
* The requirements.txt "generator" was improved to remove comments, empty lines, and sort the list of items before trying to use it (or check its md5 sum).  This allows for more actual md5 matches between projects, in-case of comments and such in the requirements file.
* A new command was added to the command-line to flush this static cache, called cleanStatic `serverless requirements cleanStatic`
* The following new options were added to this plugin...

Variable Name | Value | Description
--- | --- | ---
useStaticCache | `false/true` | Default: false.  This will enable or disable the static cache.  After a few iterations and some usage I would like to make this default: true, as this will greatly help everyone
staticCacheLocation | `<path>` | Default: `appdirectory.userCache(appName: serverless-python-requirements)`  This will allow the user to specify where the static cache should be stored.  This will be useful for people who want to persist their python requirements cache beyond a reboot, and/or for CI/CD build servers that use shared-storage to allow multiple build machines to leverage a cache to speed builds up.  For example, one example to override this to on OS-X is `~/.serverless/cache` which is serverless's working cache path.
staticCacheMaxVersions | `<integer>` | Default: 0.  This will restrict the a maximum number of caches in the cache folder.  Setting to 0 makes no maximum number of versions.  This will be useful for build/CI/CD machines that have limited disk space and don't want to (potentially) infinitely cache hundreds/thousands of versions of items in cache. Although, I would be disturbed if a project had hundreds of changes to their requirements.txt file.

## TODO
- [X] Feature Implementation
- [ ] Code Styling
- [ ] Adding Tests
- [ ] Updating README.md to inform users how to use it

### CODE REVIEW

**Comment by AndrewFarley** [comment]
For anyone wishing to try this and provide feedback change into your serverless project directory and remove the official serverless-python-requirements and install this one with...
```
rm -Rf node_modules/serverless-python-requirements
npm i github:andrewfarley/serverless-python-requirements#master
```
Then modify your serverless.yml to include the useStaticCache: true...
```
...
custom: 
  pythonRequirements:
    dockerizePip: true
    useStaticCache: true
...
```

**Comment by dschep** [comment]
I'll try and check it out soon. And fix up the linting errors, please: https://circleci.com/gh/UnitedIncome/serverless-python-requirements/345?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by AndrewFarley** [comment]
@dschep Feel free to just give this PR a brief glance, don't look too intensely yet.  Just want to know if there's something I'm doing that is fundamentally wrong, let me know so I can re-align.  Just fixed linting, but still have some test failures.

I spent a little time this evening looking into some test failures... I have one problem with one test. 
 The `py3.6 uses cache with dockerizePip option` test.  I'd like to propose removing this test, and removing mention and usage of having a pip (download) cache when used via Docker.  Let me explain why...

First, this pull request changes fundamentally how the Dockerization of this plugin works.  In the past, this plugin has been quite famous for "leaving trash" all over your project root if/when it fails for some fatal reason because we mounted the entire servicePath into Docker.  To me, that was a fundamentally flawed approach and I saw it all the time, all that Docker needs is a requirements.txt file and a temporary folder volume mount to spit its outputs into.

So, in this PR, Docker was adjusted in this way works/mounts only a dedicated sub-directory, not your entire-project-directory.  This is for security, integrity, and also for caching purposes. 

Of course, that command used in this example still works, the pipCmdExtraArgs command get sent to pip, but that .requirements-cache is made as a sub-directory under the volume mount of Docker, which, is basically a temp folder (or if using static cache, is in a cache specific to a requirements.txt folder).  Effectively making this use-case and example useless.  I see this static cache feature as superseding the existing "download-cache" in every way, especially pertaining to using docker, so maybe it's okay and I can just remove this test and the content from the README and replace with the static cache information?

I'd love thoughts/input on this.  I'd prefer to keep Docker's scope and volume mounts extremely limited and outside the full project root (servicePath) as possible, giving it only a requirements.txt file to work on and a folder to put requirements outputs.  This guarantees that Docker never "craps" onto my project root which this plugin has done to me, literally, a couple dozen times in the past (although admittedly, not recently).  I would prefer we leave it like this though, and remove mention of Dockerizing with a pip cache.

Anyways, I'll try to fix a few of the other bugs, and make all the tests pass over the next few days, and post my progress and/or update the TODO's as I go.

**Comment by Valian** [comment]
@AndrewFarley I'm using your version of serverless-python-requirements for a couple days, and it works as intended using `useStaticCache: true`. Not found time yet to test all options, though. My deployment time dropped from ~4 minutes to ~1 minute, so I'm really impressed, great job :) I'm using Ubuntu 16.04.

**Comment by AndrewFarley** [comment]
Thanks for more sets of eyes and/or trying it @Valian.  Good to hear you've seen the same results.  I'm slammed with work this week but I hope to finish up the last tests and do a few more tweaks so we can get this in mainline soon.

**Comment by cgrimal** [comment]
Hi @AndrewFarley!

I tested your PR with the `dockerSsh` (and even the `dockerFile` option) and even though it did not work at first, I managed to make it work with a minor patch.
I don't know if there is a simple way to communicate it to you, so I'll try like that as it is simple. In the block starting at line 132, replace:
```javascript
       // Use same user so requirements folder is not root and so --cache-dir works
       cmdOptions.push('-u', `${process.getuid()}`);
```
by
```javascript
      // Set the ownership of the folder to current user
      pipCmd = quote(pipCmd);
      const chownCmd = quote([
        'chown', '-R', `${process.getuid()}:${process.getgid()}`, '/var/task',
      ]);
      pipCmd = ['/bin/bash', '-c', '"' + pipCmd + ' && ' + chownCmd + '"'];
```

You'll need to add this import as well:
```javascript
const {quote} = require('shell-quote');
```

See my comment here as well: https://github.com/UnitedIncome/serverless-python-requirements/issues/182#issuecomment-393553390 as I don't know if @dschep will want to keep it like that.

Thanks!

**Comment by kichik** [comment]
@cgrimal can you please provide more details on how it fails with `-u`? I replaced `chmod` with it originally so `--cache-dir` starts working in #145.

**Comment by AndrewFarley** [comment]
Just did the merge, resolving conflicts with the latest from master, it fixed a handful of bugs that I had already sort-of coded around... I haven't gone and ran the unit tests or linting yet... working on it...

**Comment by AndrewFarley** [comment]
Hey guys, so, back at it again and think it's ready to go, rebased and refactored.  All tests are passing locally but don't seem to pass on CircleCI, I'll try to fix this soon if I can figure it out.  But if I could get @cgrimal to test your UID fix in this MR and to maybe test the DockerSSH overall?  And @Valian / @dschep could you re-review?  I refactored this quite a bit to change less code, be simpler, and work more like it used to in every way when the new features are disabled.  Besides that just need to manually spot check on Windows.  I'll wait till there's no other feedback/stoppers and do some final testing there.

**Comment by cgrimal** [comment]
Great work @AndrewFarley! 

I tested it on my current project, and it works almost like a charm! The `dockerSsh` works as expected. 

Even though I am the author of this feature, @dschep helped me quite a lot by suggesting great ideas that ended up being implemented in the PR!

Eventually, what it does is quite straightforward. It links the files used by the SSH agent from your `.ssh` directory to the `.ssh` directory of the user inside the Docker container. By default, the Docker image is `lambci/lambda:build-python3.6` and it assume the user is root (as on AWS Lambda), that's why the used `.ssh` directory is located in the `/root` folder.

If you run the Docker process as your current user, the SSH agent won't find your SSH keys because it will look for the `.ssh` directory inside the user's home directory, and it does not exist.

I remember that at first, I tried to keep the option passing the user id, but none of the workarounds I came up with manage to do the trick.

@kichik: Regarding the issue with the `--cache-dir`, I'll be glad to have a minimal example to understand how it breaks. That being said, the caching feature is already using directly this pip option, so I'm not sure it will be necessary to keep using it as an additional pip argument.

Unfortunately, I found out the `vendor` option is broken for individual package, when this option is set at function level. Indeed, you only check if `options.vendor` is set, when it can also be `f.vendor`.

A minor remark as well is that we should escape all the path to volumes, including the ones related to `dockerSsh`, as you did for the path to the cache directory.

If you want to add me as admin, I'll be glad to contribute (I don't know how much time I'll be able to put though), or I could do a PR on your branch?

Many thanks guys, I'm really looking forward to using this feature!
Thank you for reading me until the end, I hope I did not forget anything.

**Comment by AndrewFarley** [comment]
@cgrimal You raised a few interesting points...
* First, I will probably want to throw an error for people that are/will try to use --cache-dir still as a manual pip option.  I updated the readme and removed this example, but, I should also fail if you're trying to use this plugin with that.  This will help people re-align to use the now-built-in feature.  I already thought of this but totally forgot to implement this.  Any opinions?  Should I just throw a WARNING and try to proceed...?  Or should I error out?  Anyone have an opinion?
* Second, I'll do some more testing about vendor stuff with individual packages.  I don't think we had a test for this, so I'll add a new test for this, and then fix it.  I don't know if this was a problem in master, or a new problem?  Either way, I'll add a test to cover it.
* Finally, good point about escaping paths, I'll try to get around to doing this, but if you have more knowledge/time in this regard, you're welcome to add it.  I'll go add you as a contributor to my fork now.

**Comment by cgrimal** [comment]
Oh snap!
When the docker command is successful, its output is hidden, but while I was testing something, I manually ran the command, and I noticed the following warning:
```
The directory '/var/useDownloadCache/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
```

So I'm afraid we have a problem... I have to go, but I'll try to think of a solution.

**Comment by AndrewFarley** [comment]
@cgrimal That might be exactly why the CircleCI fails... if you have it failing (I don't see that error on my Mac or Ubuntu) then maybe you can debug and fix it for us?

**Comment by AndrewFarley** [comment]
@cgrimal Another idea to fix this problem in docker that randomly everyone seems to have, is to let the pip cache be build locally in the docker container, and then moved out from a local path to the mount after pip finishes.  I think this may be the best way to solve this problem, because this removed the uid problem completely.

**Comment by dschep** [comment]
Hey guys, sorry I haven't stayed on top of this discussion I'll try and catch up to review this over the weekend.

**Comment by cgrimal** [comment]
@AndrewFarley I am not sure I understand your suggestion about the pip cache folder, but what I implemented is quite simple: I use the chmod before the pip command to give ownership to root, run the pip command, and then give ownership back to the user.

What do you think?

I tried to run the tests locally, but some are failing and I'm not sure it is related... (I'm on Ubuntu by the way)

**Comment by AndrewFarley** [comment]
Thanks @cgrimal for the help with those last few things, I just did a few final things...
* I updated the README to include information about the "old" way we used to recommend doing caching, so if someone searches the page for cache-dir they will still find the new caching stanza.
* I added some code to give a warning when trying to use --cache-dir to recommend the new usage.  This warning has a URL in it which points to the project homepage to the caching anchor.
* Added a fatal error when trying to use cache-dir with docker (as that will flat out no longer work because we're not mounting the entire project inside of docker any more)
* Added a test to make sure vendor works with packaging individually (from @cgrimal complaining this didn't work, mostly because we don't have a test to cover it)

As far as I'm concerned this is ready to merge, or it's in the last few yards.  I'll spin up a Windows VM later today or tomorrow to do some final testing over there just incase.  Or if anyone else has/regularly uses Windows, would love some usage testing, especially with both caching enabled.

**Comment by felipe-augusto** [comment]
Hi, i'm trying to use your fork with gitlab. But cache seems not to be working. I don't know if you are familiar with `.gitlabci-yml`, but here's mine:

```yml
cache:
  key: "$CI_JOB_NAME"
  paths:
    - node_modules/
    - cached_requirements/

serverless_test:
  stage: deploy
  environment:
    name: production
  only:
    - master
  script:
   - mkdir -p cached_requirements
   - npm install
   - serverless deploy
```

And my `serverless.yml`:

```yml
plugins:
  - serverless-python-requirements
custom:
  pythonRequirements:
    dockerizePip: non-linux
    useDownloadCache: true
    useStaticCache: true
    cacheLocation: './cached_requirements'
    fileName: 'requirements.serverless.txt'
    zip: true
    noDeploy: [
      'boto3',
      'botocore',
      'docutils',
      'jmespath',
      'python-dateutil',
      's3transfer',
    ]

package:
  # exclude everything...
  exclude:
    - cached_requirements/**
    - '**/*'

  # include - we have to explicitly list packages with native dependencies.
  include:
    - '*.py'
```

And here's the output in the terminal:

```
> serverless deploy

Serverless: Adding Python requirements helper...
Serverless: Generated requirements from /builds/requirements.serverless.txt in /builds/.serverless/requirements.txt...
Serverless: Installing requirements from /builds/cached_requirements/80cebef5003dd2929b2152fd5f3decb5_slspyc/requirements.txt ...
Serverless: Using download cache directory /builds/cached_requirements/downloadCacheslspyc
Serverless: Zipping required Python packages...
Serverless: Packaging service...
Serverless: Excluding development dependencies...
Serverless: Removing Python requirements helper...
Serverless: Injecting required Python packages to package...
Serverless: WARNING: Function process has timeout of 60 seconds, however, it's attached to API Gateway so it's automatically limited to 30 seconds.
Serverless: Uploading CloudFormation file to S3...
Serverless: Uploading artifacts...
Serverless: Uploading service .zip file to S3 (169.13 MB)...
Serverless: Validating template...
Serverless: Updating Stack...
Serverless: Checking Stack update progress...
```

The gitlab job takes the same time, so I don't know why this is not working. Reading your description, `.serverless` folder is destroyed across gitlab jobs, maybe this is the reason why cache is not working?

Also, my image uses an alpine linux distribution.

Thanks!!

**Comment by AndrewFarley** [comment]
@felipe-augusto It looks like the plugin worked okay in general, but it doesn't look like the cache is being used in that run you copy/pasted above.  This MR has a lot of specific features with CI/CD in mind (I need these, so I made 'em).  If this is on a gitlab runner, and/or depending on how your gitlab works, if it's re-checking out the code every time, then it will be re-creating the `cached_requirements` folder every time, making it not able to actually cache.  You'll need to move that folder to somewhere which is shared between all the runners, or throughout the runs.  Because your gitlab job definition points to creating this folder, this is a large sign to me that this is not going to be a folder that is present in a future run of the job.

If your Gitlab is running jobs inside a docker container, that docker container will need to have a volume mount to the host that is a "cache" folder that can be re-used between jobs.  And that folder path is what you'll want to set `cacheLocation` to.  Does that help?  I'm using this plugin right now with Gitlab CI, Jenkins, and manually right now, all successfully.  For gitlab, I'm not using Docker for my runners so I can use a shared path, (somewhat) similarly for Jenkins I have a NFS share mounted on every node that stores the static cache so all the nodes can use it.

Something this plugin does offer (it is a subtle and unmentioned feature) that the static cache only works if it detected a successful completion from a previous run.  Such that if it fails, it won't be re-used in a future run.  This also prevents a problem with collusion incase two runners try to access and build on the same shared folder at the same time.

**Comment by AndrewFarley** [comment]
@felipe-augusto And hey, if you're still struggling... I would love more eyes, attention, and feedback to this, if I can make this _even_ better for y'all, I'd love to.  Feel free to reach out, here, or privately.  Will gladly help you.  Cheers!

**Comment by felipe-augusto** [comment]
@AndrewFarley  Thanks for the quick response and kind support, I could not answer this as fast as you could, sorry about that.

You were correct when you said that the folder is being lost between jobs, but I'm using gitlab shared runners, and I think I'm not able to mount a volume using them.

My solution to that problem, is using gitlab `cache` feature, so at the end of the job, `gitlab` does this:

```
Serverless: Removing old service versions...
Creating cache serverless_test...
node_modules/: found 10491 matching files          
cached_requirements/: found 353 matching files     
Uploading cache.zip to http://runners-cache-3-internal.gitlab.com:444/runner/project/XXXXXXX/serverless_test 
Created cache
Job succeeded
```

And at the beginning of the other job, it does:

```
Checking cache for serverless_test...
Downloading cache.zip from http://runners-cache-5-internal.gitlab.com:444/runner/project/XXXXXX/serverless_test 
Successfully extracted cache
```

For `node_modules` it works like a charm:
```
$ npm install
up to date in 2.734s
```

But, maybe I'm leaving some folder behind or some configuration that I'm not aware, because I'm only caching the `cached_requirementes` folder . I'll try to cache everything that is not committed to my repo and check if it works. I will keep you posted! 👍

**Comment by AndrewFarley** [comment]
@felipe-augusto Hey, sorry I overlooked that you were using the cache feature, I haven't tried that but it should work just fine.  I'm a little busy today, but if you're still struggling with it let me know and I'll test it on my gitlab.

**Comment by AndrewFarley** [comment]
@felipe-augusto Can you confirm that your builds are actually running from `/builds`?

**Comment by felipe-augusto** [comment]
@AndrewFarley Yes, they are running from `/builds`, the paths are below

```
Serverless: Generated requirements from /builds/felipe-augusto/project/requirements.serverless.txt in /builds/felipe-augusto/project/.serverless/requirements.txt...
Serverless: Installing requirements from /builds/felipe-augusto/project/cached_requirements/80cebef5003dd2929b2152fd5f3decb5_slspyc/requirements.txt ...
Serverless: Using download cache directory /builds/felipe-augusto/project/cached_requirements/downloadCacheslspyc
```

**Comment by felipe-augusto** [comment]
@AndrewFarley I'm still having this issue, and I tested caching everything. Tomorrow, I'll try to push a public repository to gitlab, so you can check it out too.

**Comment by AndrewFarley** [comment]
Alright, I’ll spend a little time looking into this this evening along with doing some Windows spot checks.

File: lib/pip.js [hunk 1]
```diff
@@ -194,11 +282,129 @@ function copyVendors(vendorFolder, targetFolder, serverless) {
...
+  if (funcOptions.vendor) {
+    copyVendors(funcOptions.vendor, workingReqsFolder, serverless);
+  }
+
+  // Then touch our ".completed_requirements" file so we know we can use this for static cache
+  if (options.useStaticCache) {
+    fse.closeSync(
+      fse.openSync(path.join(workingReqsFolder, '.completed_requirements'), 'w')
+    );
+  }
```
Comment (felipe-augusto) [inline lib/pip.js]: @AndrewFarley I think I know what's happening here. Your code is actually working for the `useDownloadCache`, I failed to notice it because the package I'm installing requires a lot of compiling, but I logged the docker container logs and this part is actually working. 👍 

What fails to work is `useStaticCache`, something is removing the folder that `.endWith('_slspyc')`, along with the `.completed_requirements` file that you use to verify that the static cache actually exists.

I tested this by adding `throw new Error()` here. The code fails but the folder remains!!!

image

Then I removed the error and tried again, and the cached worked! But then something erases the folder.

```sh
Serverless: Using static cache of requirements found at /Users/felipesantos/apps/sls-pip-cache/cached_requirements/ee028457a1ea41250e8af2a379765ce5_slspyc ...
Serverless: Zipping required Python packages for func1...
Serverless: Packaging service...
Serverless: Excluding development dependencies...
```

image

I think serverless is erasing it, but idk how to prevent. Do you know?

Thanks!

File: lib/pip.js [hunk 2]
```diff
@@ -211,36 +417,56 @@ function installAllRequirements() {
...
+            '.serverless',
+            `${f.module}`,
+            'requirements'
+          );
+          if (
+            reqsInstalledAt &&
+            !fse.existsSync(modulePath) &&
+            reqsInstalledAt != modulePath
+          ) {
+            if (this.options.staticCache) {
```
Comment (felipe-augusto) [inline lib/pip.js]: Found!

```
this.options.useStaticCache
```

**Comment by AndrewFarley** [comment]
@felipe-augusto I can't replicate your problem... I took your sample config and put it in my gitlab and on the second run it used the static cache perfectly...  Here's my configs, closely replicated yours as much as possible...

```yaml
service: test

provider:
  name: aws                       # Which provider to use
  runtime: python3.6              # Which Lambda runtime to use
  region: eu-west-1		  # Which AWS region to deploy in, eu-w
  memorySize: 128                 # In megabytes, 128 minimum in 64MB increments
  timeout: 60                     # In seconds

plugins:
  - serverless-python-requirements
custom:
  pythonRequirements:
    dockerizePip: non-linux
    useDownloadCache: true
    useStaticCache: true
    cacheLocation: './cached_requirements'
    fileName: 'requirements.serverless.txt'
    zip: true

package:
  # exclude everything...
  exclude:
    - cached_requirements/**
    - '**/*'

functions:
  api:
    description: All API calls routed through a single handler 
    handler: handler.lambda_handler
```

And here's my gitlab runs...
```bash
Fetching changes...
Removing .requirements.zip
Removing .serverless/
Removing cached_requirements/
Removing node_modules/
HEAD is now at 3dda851 readme
Checking out 3dda8514 as master...
Skipping Git submodules setup
Checking cache for serverless_test...
Successfully extracted cache
$ mkdir -p cached_requirements
$ npm install
$ serverless deploy
Serverless: Adding Python requirements helper...
Serverless: Generated requirements from /builds/farley/sample-serverless-project/requirements.serverless.txt in /builds/farley/sample-serverless-project/.serverless/requirements.txt...
Serverless: Installing requirements from /builds/farley/sample-serverless-project/cached_requirements/a29a14b370fd47fe2a7d1c665bcba045_slspyc/requirements.txt ...
Serverless: Using download cache directory /builds/farley/sample-serverless-project/cached_requirements/downloadCacheslspyc
Serverless: Zipping required Python packages...
```

Second run uses the static cache...
```bash
...
Fetching changes...
Removing .requirements.zip
Removing .serverless/
Removing cached_requirements/
Removing node_modules/
HEAD is now at 3dda851 readme
Checking out 3dda8514 as master...
Skipping Git submodules setup
Checking cache for serverless_test...
Successfully extracted cache
$ mkdir -p cached_requirements
$ npm install
$ serverless deploy
Serverless: Adding Python requirements helper...
Serverless: Generated requirements from /builds/farley/sample-serverless-project/requirements.serverless.txt in /builds/farley/sample-serverless-project/.serverless/requirements.txt...
Serverless: Using static cache of requirements found at /builds/farley/sample-serverless-project/cached_requirements/a29a14b370fd47fe2a7d1c665bcba045_slspyc
Serverless: Zipping required Python packages...
...
```

File: lib/pip.js (hunk 2 above)
Comment (AndrewFarley) [reply lib/pip.js]: See my comment below, I am not quite sure your problem, I tried your config on gitlab and it worked perfectly, using caching perfectly.

Comment (felipe-augusto) [reply lib/pip.js]: @AndrewFarley The problem is when you package individually, the code in my gitlab does not do that, that's why it works. The code is wrong here, you typed `staticCache` instead of `useStaticCache`.

Comment (AndrewFarley) [reply lib/pip.js]: @felipe-augusto Ahh yeah that was the old variable name... thanks for catching that!  I searched the codebase there were no other instances.  I just pushed a fix, try it now?  :)

**Comment by AndrewFarley** [comment]
I did some basic tests on Windows 10, it appears to work, so I'm calling this ready.  :P

**Comment by AndrewFarley** [comment]
FYI, The conflict is irrelevant, this merge removes the DockerPathForWin that is no longer necessary since we're not using the current folder as a path into docker, we're using already windows-friendly docker paths from appdir.

File: lib/pip.js (hunk 2 above)
Comment (felipe-augusto) [reply lib/pip.js]: @AndrewFarley Yes, I can confirm that is actually working! Thanks!

Comment (AndrewFarley) [reply lib/pip.js]: Great thanks for the feedback.  Hope it saves you tons of time on your builds like it does us.  :)

Comment (felipe-augusto) [reply lib/pip.js]: Of course it does, this needs to be merged ASAP. My build went from 22 minutes to 4. Really awesome work! 🥇

**Comment by sweepy84** [comment]
Dying for this feature - in desperate need of caching to speed things up! Will it be released soon?

**Comment by AndrewFarley** [comment]
Hey @sweepy84 I'll re-resolve conflicts shortly to help support your idea.  I'd say this is safe to merge in my opinion, I use, and have been using this branch now for months in about 5 production-level CI pipelines with 100% success, but they are either Mac/Linux (on client-side) or Linux (server side, CI runners, typically Ubuntu/Debian).  I needed this feature-set badly, so I made it.  :P

So, of quick note, this merge changes how Docker works quite a bit, it makes it more reasonable. 
 Previously this plugin mounted the entire current folder in a volume inside of Docker which was not necessary and prone to lots of odd failure regarding path escaping, uids, folder permissions, etc.  Additionally, and there's been a handful of bugs/merges recently regarding this plugin and Windows.  To my knowledge, this plugin does not currently run test coverage on Windows automatically and to my knowledge no one runs it manually (via Cygwin) if that even works.  So if you ask me, if we want to get "serious" about Windows compatibility, we have to make our CI run on Windows.  This, however, is a bit of a challenge, as to use this plugin on Windows you literally need Windows 10 Pro with Hyper-V enabled and not a VM (that I could tell) because that's what is required for Docker to work, a typical non-virtualized Windows environment.

Given that the usage on Windows is pretty thin, and people don't seem to be raising their hands to test this on Windows, I'd say this is safe to merge.  People can re-raise problems that occur, and I'll be pretty responsible to fix them.  Thoughts?  @dschep ?

**Comment by sweepy84** [comment]
@AndrewFarley - Great stuff! I'm actually running Windows so I'll be more than happy to test this out on the project I'm working on. When can we get this in?

**Comment by AndrewFarley** [comment]
@sweepy84 Could you test now, perchance?  In some of your working projects remove the node_modules/serverless-python-requirements folder and then install from my branch...
```
npm i github:andrewfarley/serverless-python-requirements#master --save
```

**Comment by sweepy84** [comment]
hey @AndrewFarley  - started testing, but didnt get very far, unfortunately.

Existing config is:
````
custom:
  pythonRequirements:
    slim: true
    slimPatterns:
      - ".*testdata.*"
      - "*.exe"
    pythonBin: python # Forced to set this as it fails to find python3.6 executable
    dockerizePip: true
````
### Test 1 - Installation Only - FAIL
1. I installed from your branch 
2. Run 'sls deploy' 
(without changing any options)

**Result** 
> Unable to find good bind path format

### Test 2 - Without Docker - SUCCESS

1. Removed the following option 
```
dockerizePip: true
```
2. Run 'sls deploy'

**Result**
It completely deployment successfully.

### Test 3 - FAIL
1. Add the feature options 
```
useDownloadCache: true
useStaticCache: true
```
2. Run 'sls deploy'

**Result**
>EPERM: operation not permitted, scandir 'C:\dev\myproject\.serverless\requirements'

After doing some research I found this [problem with npm](https://github.com/npm/npm/issues/17747) . Unfortunately, I tried clearing cache and reinstalling latest npm (v5.6) rerunning, and even using admin privileges in shell, still getting that error.

**Comment by farridav** [comment]
*Bump I'm also very reliant on this plugin, and have super slow build times, Id love to see this rebased (again) and merged in ...

**Comment by AndrewFarley** [comment]
@sweepy84  The results were helpful, if only to tell me that I (or someone) needs to get a Windows machine and help this merge request work properly in that environment.

@farridav I've spent a lot of time rebasing this keeping this working with the latest code, but I don't always/continue to have time rebasing this code.  This code is in place now in 4 pipelines working very successfully, but all of them are non-windows environments (Linux/OS-X), so this already fixed my problem.  I will try to continue to rebase when I have a few moments.

Bottom line is, I, or someone, needs to get a Windows machine (or many of them) to make this work properly on Widnows... or maybe we make Windows support optional for this feature-set?  Unfortunately, due to the nature of Docker on Windows, I can't just spin up a VM on the cloud to test this, it needs a raw/real machine, and the highest level of Windows to do this.  I neither have the time nor the money to waste nor the machines laying around to do this.  Anyone?

**Comment by sweepy84** [comment]
@dschep  - I'm actually getting a different error now

> Unable to find good bind path format

Using:

pythonRequirements:
    slim: true
    slimPatterns:
    - "**/*test*"
    - "*.exe"
    - "**/*.txt"
    pythonBin: python # Forced to set this as it fails to find python3.6 executable
    useDownloadCache: true
    useStaticCache: true
    dockerizePip: true or non-linux`

**Comment by dee-me-tree-or-love** [comment]
@sweepy84 I had been running into the bind path problem on Windows as well - for me the issue was with the [volume mounting for the `docker-machine`](https://docs.docker.com/machine/reference/mount/) - so I just added a path binding option that would satisfy my settings. You can see the path that are attempted in [_`docker.js : L122`_](https://github.com/UnitedIncome/serverless-python-requirements/blob/d00d095406cfe5845ada740efe03a3dca0190264/lib/docker.js#L122). 

Probably not a solution, but might help 🙃

**Comment by AndrewFarley** [comment]
Yeah, I worried that might be the case, hmph.  Alright, I think I know how to fix it... but I need to get Docker on Windows working.  Do any of you have a throwaway/temporary machine I can RDP into to try to debug this?  Would save me unnecessary cost of buying Windows Ultimate just to get this feature implemented.  @dee-me-tree-or-love  / @sweepy84 , or again, maybe someone can step up with Windows and help me fix this in that platform?  Let me know.

My goal is/was never to alienate anyone @sweepy84, I just lack the resources to properly test/iterate on that platform.  Complain to Microsoft not allowing docker on anything except the most expensive version of windows, and not on a VM as well.  :(  I tried a few VM providers (aws, gc) docker's not possible on the cloud far as I can tell.  If you know how/where, point me in that direction and I'll make it work.

**Comment by kichik** [comment]
[Docker Toolbox](https://docs.docker.com/toolbox/toolbox_install_windows/) is still available and works without Windows 10 Pro and without Hyper-V.

**Comment by AndrewFarley** [comment]
Hmm, I'll try it out and let you guys know, thanks @kichik

**Comment by dschep** [comment]
Sorry I can't really provide any help on bindpath issues on windows @sweepy84. I didn't write that feature and don't have the hardware or software licenses to test it :confused:

**Comment by AndrewFarley** [comment]
Alright ladies and gentleman... I've got Windows support fully working.  99% of the way there... @sweepy84 / @kichik / @dee-me-tree-or-love or anyone else with active Windows codebases can you try this?  I tested/confirmed all three scenarios work that @sweepy84 reported was failing above.  Without docker works, with docker works, with static caching works and it uses the static cache.

All tests are passing, all my active pipelines work with this code (tested against 5 different codebases/pipelines), and I can confirm all the things seem to work fine with Docker Toolbox on Windows 10.

Thanks for any testing/attention, sorry this one has taken forever, since it "works for me" I really lost interest in trying to get it merged (and I couldn't justify working on it on a clients dime because of that).  Hope we're close and can get everyone able to use this soon.  And in the near future have both download and static caching enabled by default (because it's amazing).  Cheers!

**Comment by AndrewFarley** [comment]
Thought I'd share a success screenshot.  Without caching enabled, this codebase takes ~2.5min for the package to build, with caching around 15 seconds.  :P
screen shot 2018-09-02 at 2 23 58 am

**Comment by farridav** [comment]
Nice ! @dschep / @Valian any chance of an approve and merge ?

**Comment by sweepy84** [comment]
YESSSS!! @AndrewFarley  - great job mate! Looks like its working well! oh the time we're going to save!

I saw your edit about not caching in windows only - what do u mean by "Windows only" - looks to be working for me :)

**Comment by AndrewFarley** [comment]
Re: @sweepy84 there is two kinds of caching, the "static" cache, or the output of pip, and the "download" cache, the intermediary downloads that pip downloads to feed into pip.  The static cache is working perfectly on windows, but the download cache doesn't appear to work.  But as I said, it's not doing this in a fatal way, it's just keeping the cache in docker from what I can tell.  I don't have any more time this weekend to spend on this, I can try again next weekend to fix this feature, or we can merge it and I can fix that in a future patch.

**Comment by sweepy84** [comment]
@AndrewFarley - oh ok - thanks clarifying. 

My 2 cents - makes sense for it to be merged in and fixed at a later time (ideally raise the bug after doing the merge so it's not forgotten) given it's size so you don't waste time rebasing.

**Comment by dschep** [comment]
Good job on all the hardwork @AndrewFarley, I'll get to reviewing this ASAP. Sometime this week.

**Comment by dschep** [comment]
I'll publish to NPM on monday when i'm on a computer with npm creds

**Comment by dschep** [comment]
Released as v4.2.0!!
